src/pykeen/nn/functional.py

```diff
@@ -1069,53 +1069,30 @@ def pair_re_interaction(
     )
 
 
-def _rotate_quaternion(qa: torch.FloatTensor, qb: torch.FloatTensor) -> torch.FloatTensor:
-    # Rotate (=Hamilton product in quaternion space).
-    return torch.cat(
-        [
-            qa[0] * qb[0] - qa[1] * qb[1] - qa[2] * qb[2] - qa[3] * qb[3],
-            qa[0] * qb[1] + qa[1] * qb[0] + qa[2] * qb[3] - qa[3] * qb[2],
-            qa[0] * qb[2] - qa[1] * qb[3] + qa[2] * qb[0] + qa[3] * qb[1],
-            qa[0] * qb[3] + qa[1] * qb[2] - qa[2] * qb[1] + qa[3] * qb[0],
-        ],
-        dim=-1,
-    )
-
-
-def _split_quaternion(x: torch.FloatTensor) -> torch.FloatTensor:
-    return torch.chunk(x, chunks=4, dim=-1)
-
-
 def quat_e_interaction(
     h: torch.FloatTensor,
     r: torch.FloatTensor,
     t: torch.FloatTensor,
+    table: torch.FloatTensor,
 ):
     """Evaluate the interaction function of QuatE for given embeddings.
 
     The embeddings have to be in a broadcastable shape.
 
-    .. note ::
-        dim has to be divisible by 4.
-
-    :param h: shape: (`*batch_dims`, dim)
+    :param h: shape: (`*batch_dims`, dim, 4)
         The head representations.
-    :param r: shape: (`*batch_dims`, dim)
+    :param r: shape: (`*batch_dims`, dim, 4)
         The head representations.
-    :param t: shape: (`*batch_dims`, dim)
+    :param t: shape: (`*batch_dims`, dim, 4)
         The tail representations.
+    :param table:
+        the quaternion multiplication table.
 
     :return: shape: (...)
         The scores.
     """
-    return -(
-        # Rotation in quaternion space
-        _rotate_quaternion(
-            _split_quaternion(h),
-            _split_quaternion(r),
-        )
-        * t
-    ).sum(dim=-1)
+    # TODO: this sign is in the official code, too, but why do we need it?
+    return -torch.einsum("...di, ...dj, ...dk, ijk -> ...", h, r, t, table)
```

Comment:
Can we just call the function to get the table inside here rather than passing it? Still I don’t see why we would ever use anything else besides the exact one from
That function

Reply:
The tensor has to be on the same device as the `h, r, t` for this code to work. Also it has to be of suitable data type (in case you are not using the default float, but, e.g., are using mixed-precision).

By having the tensor be a buffer of the interaction function, which derives from `nn.Module`, torch / PyL will take care of it.

Reply:
Good enough for me. Ship it!

Reply:
After tests pass ;)

Reply:
Blocked by https://github.com/pykeen/pykeen/pull/1058

Reply:
can we switch this now?

Reply:
yes, cf. https://github.com/pykeen/pykeen/pull/1056/commits/81296e199c7bf6d45fdc3c0741fec376d9854b0c